lib/spack/spack/concretize.py

```diff
@@ -465,6 +472,54 @@ def _compiler_concretization_failure(compiler_spec, arch):
         raise UnavailableCompilerVersionError(compiler_spec, arch)
 
 
+def concretize_specs_together(*abstract_specs):
+    """Given a number of specs as input, tries to concretize them together.
+
+    Args:
+        *abstract_specs: abstract specs to be concretized
+
+    Returns:
+        List of concretized specs
+    """
+    # Create a temporary repository with an umbrella package
+    # that depends on each spec
+    tmpdir = tempfile.mkdtemp()
+    repo_path, namespace = spack.repo.create_repo(tmpdir)
+
+    debug_msg = '[CONCRETIZATION]: Creating helper repository in {0}'
+    tty.debug(debug_msg.format(repo_path))
+    pkg_dir = os.path.join(repo_path, 'packages', 'concretizationroot')
+    fs.mkdirp(pkg_dir)
+
+    environment = spack.tengine.make_environment()
+    template = environment.get_template('misc/coconcretization.pyt')
+
+    # Split recursive specs, as it seems the concretizer has issue
+    # respecting conditions on dependents expressed like
+    # depends_on('foo ^bar@1.0')
+    split_specs = []
+    for spec in abstract_specs:
+        split_specs.extend([s.strip() for s in spec.split('^')])
+
+    abstract_specs = [spack.spec.Spec(s) for s in abstract_specs]
```

Comment:
Can you move this line to the top of the method? I think that will improve clarity, since this line is effectively input sanitization.

Reply:
It was at the top of the method, but moved here because it was more convenient to operate directly on strings to obtain `split_specs`. See #11160 as for why `split_specs` was necessary.

Reply:
Will try to address this together with refactoring the creation of the concretization repository into its own function.